src/components/popup/popup.tsx

```diff
@@ -0,0 +1,128 @@
+            >
+                {this.props.children}
+            </div>
+        );
+    }
+
+    private setPosition = (popup: any) => {
+        if (popup) {
+            const newStyle = this.state.style;
+            const popupRect = popup.getBoundingClientRect();
+            let popupWidth = popupRect.width;
+
+            newStyle.maxHeight = this.props.maxHeight;
+            if (this.props.syncWidth) {
+                newStyle.width = (this.props.anchor as HTMLElement).getBoundingClientRect().width;
+                popupWidth = newStyle.width;
+            }
+
+            const anchorHorizontalPos = this.props.anchorPosition!.horizontal;
+            const anchorVerticalPos = this.props.anchorPosition!.vertical;
+
+            setTop(newStyle, (this.props.anchor as HTMLElement).getBoundingClientRect(),
+                anchorVerticalPos, popup.getBoundingClientRect().height, this.props.popupPosition!.vertical);
+            setLeft(newStyle, (this.props.anchor as HTMLElement).getBoundingClientRect(),
+                anchorHorizontalPos, popupWidth, this.props.popupPosition!.horizontal);
+            if (this.props.syncWidth) {
+                newStyle.width = (this.props.anchor as HTMLElement).getBoundingClientRect().width;
+            }
+            this.setState({style: newStyle});
+        }
+    }
+}
+
+function setTop(popupStyle: CSSProperties, anchorRect: ClientRect,
+                anchorPosition: VerticalPosition, popupHeight: number, popupPoistion: VerticalPosition) {
+    switch (popupPoistion) {
+        case 'top':
+            popupStyle.top = getVerticalReference(anchorRect, anchorPosition);
+            break;
+        case 'center':
+            popupStyle.top = getVerticalReference(anchorRect, anchorPosition) - (popupHeight / 2);
+            break;
+        case 'bottom':
+            popupStyle.bottom = `calc(100% - ${getVerticalReference(anchorRect, anchorPosition)}px`;
+            break;
+    }
+}
+
+function setLeft(popupStyle: CSSProperties, anchorRect: ClientRect,
+                 anchorPosition: HorizontalPosition, popupWidth: number, popupPoistion: HorizontalPosition) {
+    switch (popupPoistion) {
+        case 'left':
+            popupStyle.left = getHorizontalReference(anchorRect, anchorPosition);
+            break;
+        case 'center':
+            popupStyle.left = getHorizontalReference(anchorRect, anchorPosition) - (popupWidth / 2);
+            break;
+        case 'right':
+            popupStyle.right = `calc(100% - ${getHorizontalReference(anchorRect, anchorPosition)}px`;
+            break;
+    }
+}
+
+function getVerticalReference(rect: ClientRect, anchorPosition: VerticalPosition) {
+    if (anchorPosition === 'center') {
+        return (window.scrollY ? window.scrollY : window.pageYOffset) + rect.top + (rect.height / 2);
+    } else {
+        return (window.scrollY ? window.scrollY : window.pageYOffset) + rect[anchorPosition as 'top' | 'bottom'];
```

Comment:
you can calculate and save `window.scrollY ? window.scrollY : window.pageYOffset` before `if`